src/elliptic.jl

```diff
@@ -0,0 +1,244 @@
+# References:
+#  [1] Abramowitz, Stegun: Handbook of Mathematical Functions (1965)
+#  [2] ellipjc.m in Toby Driscoll's Schwarz-Christoffel Toolbox
+
+
+#------------------------------------------------
+# Descending and Ascending Landen Transformation
+
+descstep(m) = m/(1+sqrt(1-m))^2
+
+@generated function shrinkm(m,::Val{N}) where {N}
+    # [1, Sec 16.12] / https://dlmf.nist.gov/22.7.i
+    quote
+        f = one(m)
+        Base.Cartesian.@nexprs $N i->begin
+            k_i = descstep(m)
+            m = k_i^2
+            f *= 1+k_i
+        end
+        return (Base.Cartesian.@ntuple $N k), f, m
+    end
+end
+
+function ellipj_smallm(u,m)
+    # [1, Sec 16.13] / https://dlmf.nist.gov/22.10.ii
+    if VERSION < v"0.7-"
+        sinu = sin(u)
+        cosu = cos(u)
+    else
+        sinu, cosu = sincos(u)
+    end
+    sn = sinu - m*(u-sinu*cosu)*cosu/4
+    cn = cosu + m*(u-sinu*cosu)*sinu/4
+    dn = 1 - m*sinu^2/2;
+    return sn,cn,dn
+end
+function ellipj_largem(u,m1)
+    # [1, Sec 16.15] / https://dlmf.nist.gov/22.10.ii
+    sinhu = sinh(u)
+    coshu = cosh(u)
+    tanhu = tanh(u)
+    sechu = sech(u)
+    sn = tanhu + m1*(sinhu*coshu-u)*sechu^2/4
+    cn = sechu - m1*(sinhu*coshu-u)*tanhu*sechu/4
+    dn = sechu + m1*(sinhu*coshu+u)*tanhu*sechu/4
+    return sn,cn,dn
+end
+
+@generated function ellipj_growm(sn,cn,dn, k::NTuple{N,<:Any}) where {N}
+    # [1, Sec 16.12] / https://dlmf.nist.gov/22.7.i
+    quote
+        Base.Cartesian.@nexprs $N i->begin
+            kk = k[end-i+1]
+            sn,cn,dn = (1+kk)*sn/(1+kk*sn^2),
+                           cn*dn/(1+kk*sn^2),
+                     (1-kk*sn^2)/(1+kk*sn^2)
+                     # ^ Use [1, 16.9.1]. Idea taken from [2]
+        end
+        return sn,cn,dn
+    end
+end
+@generated function ellipj_shrinkm(sn,cn,dn, k::NTuple{N,<:Any}) where {N}
+    # [1, Sec 16.14] / https://dlmf.nist.gov/22.7.ii
+    quote
+        Base.Cartesian.@nexprs $N i->begin
+            kk = k[end-i+1]
+            sn,cn,dn = (1+kk)*sn*cn/dn,
+                     (cn^2-kk*sn^2)/dn, # Use [1, 16.9.1]
+                     (cn^2+kk*sn^2)/dn  # Use [1, 16.9.1]
+        end
+        return sn,cn,dn
+    end
+end
+
+function ellipj_viasmallm(u,m,::Val{N}) where {N}
+    k,f,m = shrinkm(m,Val{N}())
+    sn,cn,dn = ellipj_smallm(u/f,m)
+    return ellipj_growm(sn,cn,dn,k)
+end
+function ellipj_vialargem(u,m,::Val{N}) where {N}
+    k,f,m1 = shrinkm(1-m,Val{N}())
+    sn,cn,dn = ellipj_largem(u/f,m1)
+    return ellipj_shrinkm(sn,cn,dn,k)
+end
+
+
+#----------------
+# Pick algorithm
+
+Base.@pure puresqrt(x) = sqrt(x)
```

Comment:
Isn't `sqrt` already pure?

Reply:
Actually, it isn't for `BigFloat`, but that is intentional, since we can change precision via `setprecision`.

Reply:
It also isn't for the `FloatXX` types:
```
julia> foo(T) = Val{sqrt(eps(T))}()
foo (generic function with 1 method)

julia> @code_warntype foo(Float64)
Variables:
  #self# <optimized out>
  T <optimized out>

Body:
  begin 
      return ((Core.apply_type)(Main.Val, (Base.Math.sqrt_llvm)(2.220446049250313e-16)::Float64)::Type{Val{_}} where _)()::Val{_} where _
  end::Val{_} where _

julia> foo(T) = Val{puresqrt(eps(T))}()
foo (generic function with 1 method)

julia> @code_warntype foo(Float64)
Variables:
  #self# <optimized out>
  T <optimized out>

Body:
  begin 
      return $(QuoteNode(Val{1.4901161193847656e-8}()))
  end::Val{1.4901161193847656e-8}
```
But I agree, I should have done 
```
puresqrt(x) = sqrt(x)
Base.@pure puresqrt(x::Union{Float16,Float32,Float64}) = sqrt(x)
```
With this, I get
```
julia> foo(T) = Val{puresqrt(eps(T))}()
foo (generic function with 1 method)

julia> @code_warntype foo(Float64)
Variables:
  #self# <optimized out>
  T <optimized out>

Body:
  begin 
      return $(QuoteNode(Val{1.4901161193847656e-8}()))
  end::Val{1.4901161193847656e-8}

julia> @code_warntype foo(BigFloat)
Variables:
  #self# <optimized out>
  T::Type{BigFloat}

Body:
  begin 
      SSAValue(0) = $(Expr(:invoke, MethodInstance for eps(::Type{BigFloat}), :(Main.eps), :(T)))
      return ((Core.apply_type)(Main.Val, $(Expr(:invoke, MethodInstance for sqrt(::BigFloat), :(Main.sqrt), SSAValue(0))))::Type{Val{_}} where _)()::Val{_} where _
  end::Val{_} where _
```
Just out of curiosity, is there any fundamental obstacle to do this in Base, and possible for all numerical functions?

Reply:
Actually, maybe it would be better to just do 
```
Base.@pure puresqrt(x) = Float64(sqrt(x))
```
Then `puresqrt` is pure, and since the `nsteps` computations are done in `Float64` anyway, I don't lose anything in terms of generality.